File: m4-h2/src/main/java/ru/yegorpilipenko/otus/spring/m4h2/service/AuthorServiceImpl.java
package ru.yegorpilipenko.otus.spring.m4h2.service;

import com.netflix.hystrix.contrib.javanica.annotation.HystrixCommand;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import ru.yegorpilipenko.otus.spring.m4h2.model.Author;
import ru.yegorpilipenko.otus.spring.m4h2.repository.BookRepository;

import java.util.Collection;

@Service
@RequiredArgsConstructor
public class AuthorServiceImpl implements AuthorService {

    private final BookRepository bookRepository;

    @Override
    @HystrixCommand(groupKey = "Authors", commandKey = "GetAllAuthors")
    public Collection<Author> getAll() {
        return bookRepository.findAllAuthors();
    }

}

File: m2-h3/src/main/kotlin/ru/yegorpilipenko/otus/spring/m2h3/service/AuthorServiceImpl.kt
package ru.yegorpilipenko.otus.spring.m2h3.service

import org.springframework.stereotype.Service
import ru.yegorpilipenko.otus.spring.m2h3.model.Author
import ru.yegorpilipenko.otus.spring.m2h3.model.Book
import ru.yegorpilipenko.otus.spring.m2h3.repository.BookRepository

@Service
class AuthorServiceImpl(val bookRepository: BookRepository): AuthorService {

    override fun getAll(): Collection<Author> = bookRepository.findAll().flatMap(Book::authors).toSet()

}
